Add NWS weather

MatthewFlamm opened the pull request:
## Description:
Adds NWS weather platform.  [NWS weather](https://www.weather.gov).

There are two forecast modes supported by NWS, one is a separate day and night forecast (two forecasts a day) or hourly.  Hourly works well with the standard card.  Day and night displays the hour of the forecast which is confusing.  This platform has an additional forecast attribute to tag which forecast is day or night.   A precipitation probability is also exposed.  A modified weather card PR in the frontend is planned to utilize these features to better support day and night forecasts.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** https://github.com/home-assistant/home-assistant.io/pull/9372

## Example entry for `configuration.yaml` (if applicable):
```yaml
weather:
  - platform: nws
    api_key: my_email
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly ([example][ex-manifest]).
  - [x] New dependencies have been added to `requirements` in the manifest ([example][ex-requir]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-manifest]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json
[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json#L5
[manifest-docs]: https://developers.home-assistant.io/docs/en/development_checklist.html#_the-manifest-file_

MatthewFlamm committed:
Add nws weather.

MatthewFlamm committed:
Hassfest

fabaff commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,306 @@
+    """
+    Parse icon url to NWS weather codes.
+
+    Example:
```
Remove example as it's not working aka gives a HTTP 404 error.

MatthewFlamm replied:
The example was indeed wrong, and it is now fixed.  I do think it adds legibility to the code, but let me know if you'd still prefer it to be removed

MatthewFlamm replied:
I also included a link to all the icon weather codes above CONDITION_CLASSES, instead of just mentioning only the one code that didn't map.  There are others like `dust`.

fabaff commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,306 @@
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Set up the nws platform."""
```
```suggestion
    """Set up the MWS weather platform."""
```

fabaff commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,306 @@
+    latitude = config.get(CONF_LATITUDE, hass.config.latitude)
+    longitude = config.get(CONF_LONGITUDE, hass.config.longitude)
+    station = config.get(CONF_STATION)
+    api_key = config.get(CONF_API_KEY)
```
Use
```suggestion
    api_key = config[CONF_API_KEY]
```
as this key has to be present.

fabaff commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,306 @@
+        _LOGGER.error("Latitude/longitude not set in Home Assistant config")
+        return
+
+    from pynws import Nws
```
Move it up to the start of the function.

fabaff commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,306 @@
+
+    websession = async_get_clientsession(hass)
+    # ID request as being from HA, pynws prepends the api_key in addition
+    api_key_ha = [api_key + 'homeassistant']
```
Use string formatting instead of concatenation.

fabaff commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,306 @@
+    if station == '':
+        with async_timeout.timeout(10, loop=hass.loop):
+            stations = await nws.stations()
+        _LOGGER.info("Station list: %s", stations)
```
Set this to debug too.

fabaff commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,306 @@
+                self._forecast = await self._nws.forecast()
+            elif self._mode == 'hourly':
+                self._forecast = await self._nws.forecast_hourly()
+            else:
```
Will never happen as the configuration needs to pass the validation before the integration is set up.

fabaff reviewed (commented):
Just some comments to get the review started.

MatthewFlamm committed:
Address multiple comments

MatthewFlamm committed:
Add NWS icon weather code link

MatthewFlamm commented:
I'm going to remove the TODO from the PR. My frontend dev environment is totally hosed, and I'm not sure how to fix it.  My frontend changes did work several versions ago, but I can't test that it works today.  I don't want to bog down this PR with these issues.

MatthewFlamm commented:
#23875 is a similar PR for a NWS/NOAA sensor platform, which uses the same data source and library.  These should be considered together, as we have a few inconsistencies across the two implementations.  I'm proposing that we work together to resolve the inconsistencies, unless there is a preference to implement the weather platform or the sensor platform first.  Then we can resolve the differences during the second PR.

MatthewFlamm committed:
Add metar fallback.

Use metar code from nws observation if normal api is missing data.

MatthewFlamm committed:
only get 1 observation - we dont use more than 1

MatthewFlamm commented:
I have added a fall back to parsing the metar code if the regular data is missing.  This is to be consistent with #23875, as well as fixes issues with some stations having no formatted data but valid metar codes sometimes.  I also changed it to only retrieve the latest observation since we aren't using prior observations.

The biggest mismatch in approach I see now with #23875 , is with obtaining the condition to convert to HA.  This PR uses the icon url which has a published catalog of possible values and meanings.  The sensor PR parses a text description.

rewsssiestedbo commented:
> I have added a fall back to parsing the metar code if the regular data is missing. This is to be consistent with #23875, as well as fixes issues with some stations having no formatted data but valid metar codes sometimes. I also changed it to only retrieve the latest observation since we aren't using prior observations.
> 
> The biggest mismatch in approach I see now with #23875 , is with obtaining the condition to convert to HA. This PR uses the icon url which has a published catalog of possible values and meanings. The sensor PR parses a text description.

It is probably best if I just remove the code to translate the NWS textDescription values to HA condition codes.   None of the other weather sensors use the HA condition codes, and unlike for the weather platforms, there is nothing that depends upon the specific HA condition codes.  I had mainly put it in as a basis for a future weather platform module, but since you have that module completed, there is no reason to keep it.

MatthewFlamm committed:
add mocked metar for tests

MatthewFlamm committed:
lint

MatthewFlamm committed:
mock metar package for all tests

rewsssiestedbo commented:
The code doesn't handle the exceptions that can be raised from the pynws methods:  if the longitude/latitude is outside the area covered by the NWS, the API will return an HTTP 404 error, which pynws turns into a ClientReponseError exception.  The same thing happens if the station code is not valid.  These should be turned into a ConfigEntryNotReady exception.   Other types of errors from the pynws methods typically indicate an issue with the connection to the API (or between the Akamai edge server and the NWS backend), and should turn into a PlatformNotReady (on any call during the initial setup), or treated as a missed poll during normal updates.

rewsssiestedbo commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,329 @@
+        wind_m_s = self._observation['windSpeed']['value']
+        if wind_m_s is None and self._metar_obs:
+            wind_m_s = self._metar_obs.wind_speed.value(units='MPS')
+            print(wind_m_s)
```
This print statement should be removed - I assume it is from debugging.

rewsssiestedbo commented:
I starting running this component yesterday.  Today it logged a number of errors:
File "/home/homeassistant/.homeassistant/custom_components/nws/weather.py", line 268, in wind_bearing
wind_bearing = self._metar_obs.wind_dir.value()
AttributeError: 'NoneType' object has no attribute 'value'

This appears to be a result of the wind direction value in the JSON being null, and the field in the METAR data saying VRB (variable), which results in no value from the parse metar record either.   An example of is this the observation data for KPHL at "https://api.weather.gov/stations/KPHL/observations/2019-05-23T20:02:00+00:00"

MatthewFlamm committed:
add check for metar attributes

MatthewFlamm commented:
> AttributeError: 'NoneType' object has no attribute 'value'

This should be fixed now.  I'll work on handling errors especially during platform setup as these seem to be fatal.

MatthewFlamm committed:
catch errors in setup

MatthewFlamm committed:
add timeout error

MatthewFlamm committed:
handle request exceptions

MatthewFlamm committed:
check and test for missing observations

MatthewFlamm committed:
refactor to new pynws

MatthewFlamm commented:
@fabaff I feel that the code here has ballooned too much.  I'm working on pulling a lot of the messy logic into the upstream pypi library to provide a cleaner interface.  This is almost done.  Once that is complete, should I push here, or open a new PR?

MartinHjelmare commented:
I would say, continue here, thanks!

MartinHjelmare commented:
If it's going to take a while until the new changes are ready, please close the PR, then reopen it when you're ready for a review. We're trying to decrease our open PR buffer. Thanks!

MatthewFlamm commented:
Thanks, Changes are very close to being finished, I'll continue here

MatthewFlamm committed:
change to simpler api

MatthewFlamm commented:
In eager anticipation of dropping 3.5 support, I used an f string.  It breaks CI since I assume it is still using 3.5.  Is the preference to conform to 3.5 compatibility until _after_ the next update?  Or should I keep as is?

MartinHjelmare commented:
I suggest keeping Python 3.5 compatability until CI is updated. Otherwise we can't merge.

MatthewFlamm committed:
Make py3.5 compatible

Remove f string

MatthewFlamm commented:
FYI to reviewers, the change to the base components/weather/__init__.py file is to enable situations in which temperature is `None`.  Currently there is a TypeError due to an attempted unit conversion.

MatthewFlamm committed:
bump pynws version

MatthewFlamm committed:
gen_requirements

MatthewFlamm committed:
fix wind bearing observation

MatthewFlamm committed:
get uptodate-black

MatthewFlamm committed:
Revert "Make py3.5 compatible"

This reverts commit 4946d91779a6e539ea43e667b2265557a49a0bb5.

MatthewFlamm committed:
Precommit black missed a file?

MatthewFlamm committed:
black test

MatthewFlamm committed:
add exceptional weather condition

MatthewFlamm commented:
Given that the last real review of this PR occurred more than 3 months ago, is this just being left to die?  That is fine if this integration is unwelcome for whatever reason, but it would be nice to know whether I should continue committing my time.  I have kept it alive through some major structural changes including hassfest, great migration, and black formatting, and I'm not sure I want to keep continuing if it is destined for a slow death into obscurity.

I guess I will change my approach, and start asking questions:

1. I see several types of weather platform styles.  Some only do something like `return api_obj.data.get('key')` inside the properties.  Some also do unit conversions and condition mapping.  This integration leans toward the second implementation.  In this way, the upstream library is very faithful to the NWS API data format, and all the custom handling is done inside HA.  What is the preferred method?  Should I move ALL logic into the pypi library? This would make the pypi library very HA specific which doesn't seem desirable either IMO.

2. Similarly, I have unit conversions being done in HA code since HA has unit conversion tools.  This would be annoying to implement in the upstream library without relying on another dependency like `pint`.

3. Should I just delete the tests and put this integration into  .coveragerc? I stumbled my way through mocking the class from the pypi library, so this might be the weakest link of the PR.  I welcome any pointers or suggestions in general.

4. I'm looking into data entry flows, but I need to do more research on this.  I'm trying to figure out how to make API calls from within the flow, as the optimal flow would be first to have user input lat/lon, and then have the option of a selectable list of nearby stations that are returned from an API call.  Bonus for giving links to the NWS page for each station, so the user can investigate before choosing.  I suppose I could implement the easier method first of just replicating the current operation (the user must know what station to use or accept the closest station) inside the data entry flow methods.  This still requires an API call during the data entry flow to retrieve the closest station.  Would either of these move this PR forward?

MartinHjelmare commented on tests/components/nws/test_nws.py:
```diff
@@ -0,0 +1,337 @@
+        return
+
+
+class TestNWS(unittest.TestCase):
```
All new tests must be standalone pytest test functions. Ie not unittest class based test methods.

MatthewFlamm replied:
I had based this off of one of the existing weather platforms, but this is useful feedback.  If I cannot reconfigure the tests in a timely manner, I will remove them and plan on adding them in a separate PR.

MartinHjelmare reviewed (commented):
This is high up on my PR review todo-list, but it's hard to find time for this one. As I understand it we should look at this together with https://github.com/home-assistant/home-assistant/pull/23875. So it's around 2k  - 3k lines of code in total to digest.

1. If there's a model for the entity that requires a certain unit, we have to follow the model. Then it's ok to do conversion in home assistant.

2. See 1.

3. Tests are appreciated but not required for new integrations that interacts with 3rd party services or devices.

4. Look at some integrations that you have access to with config flow, to see how they have solved things.

MatthewFlamm commented:
Thanks.  My original plan was to make a PR for weather platform first, then a sensor platform to follow.  But then PRs started getting rejected for sensor platforms for weather based on https://github.com/home-assistant/architecture/issues/107.  See #19500.  Otherwise I would have made a PR for both weather and sensor platforms.  But, recently other sensor platforms were accepted.  See #21110.  Very confusing to me.

Coincidentally, someone took the pypi library I made and created their own sensor platform PR days after I submitted this one.  The highest form of flattery I suppose, but it certainly has made this more complicated.

My 100% biased persepective is that the sensor PR needs a lot of work, even if it is okay to accept given the confusion on weather/sensor platforms.

I have tried to be very diplomatic and upfront with the PR #23875, keeping its author up to date in changes in the upstream pypi library, which I authored.  IMHO, that PR is now out of date and includes way too much specific code into HA codebase.  I have used some of that PR as inspiration for making the upstream library more robust to missing data, but we have taken a completely different approach.  

#23875 has a lot of logic based on undocumented assumptions on when certain data is available or when to use data with certain quality codes from NWS.  My pypi library simply looks for missing data in the 'normal API', e.g. value for temperature, and then replaces with the metar coded data (which is also returned in the API), if available.  That is more robust logic IMHO, and requires no additional code in HA.  #23875 has not conformed to that transition in the upstream library yet.  I would argue that if documentation exists for how to handle the quality codes, that this logic be in the pypi library anyway.

I also don't want to step on HA dev toes by making very strong statements about these things in that PR especially if my own code hasn't been reviewed.

MartinHjelmare commented:
Thanks for this info. It's good to have when we start reviewing further.

rewsssiestedbo commented:
> Thanks. My original plan was to make a PR for weather platform first, then a sensor platform to follow. But then PRs started getting rejected for sensor platforms for weather based on [home-assistant/architecture#107](https://github.com/home-assistant/architecture/issues/107). See #19500. Otherwise I would have made a PR for both weather and sensor platforms. But, recently other sensor platforms were accepted. See #21110. Very confusing to me.
> 
> Coincidentally, someone took the pypi library I made and created their own sensor platform PR days after I submitted this one. The highest form of flattery I suppose, but it certainly has made this more complicated.
> 
> My 100% biased persepective is that the sensor PR needs a lot of work, even if it is okay to accept given the confusion on weather/sensor platforms.
> 
> I have tried to be very diplomatic and upfront with the PR #23875, keeping its author up to date in changes in the upstream pypi library, which I authored. IMHO, that PR is now out of date and includes way too much specific code into HA codebase. I have used some of that PR as inspiration for making the upstream library more robust to missing data, but we have taken a completely different approach.
> 
> #23875 has a lot of logic based on undocumented assumptions on when certain data is available or when to use data with certain quality codes from NWS. My pypi library simply looks for missing data in the 'normal API', e.g. value for temperature, and then replaces with the metar coded data (which is also returned in the API), if available. That is more robust logic IMHO, and requires no additional code in HA. #23875 has not conformed to that transition in the upstream library yet. I would argue that if documentation exists for how to handle the quality codes, that this logic be in the pypi library anyway.
> 
> I also don't want to step on HA dev toes by making very strong statements about these things in that PR especially if my own code hasn't been reviewed.

Yes, #23875 has not been updated to deal with the new approach in pynws, partly due to time.  I also wanted to see if the existing code in 23875 would encounter any more strange data issues in the API responses that might need to be address.   I agree that quite a bit of the current code is to deal with undocumented behavior of the API - unfortunately there is little documentation on the observations responses in the API documentation and from reviewing responses from a number of different types of stations over a period of a number of months it became clear that the only way to make the code robust to the various idiosyncrasies see in the data is to have some complex case handling.  

Note, I would view what I have in place is code based on documented assumptions about undocumented responses, not undocumented assumptions :)  One of the key challenges with the observation data is that when the value attribute is "null", it can mean either that there is just no update of the value in this record (and therefore should be treated as if the record did not include that observation), or that the value of the observation is actually undefined.  If every case where the value is "null" is treated as undefined, the history recorded by HA becomes at best confusing and potentially useless: some stations will only report a temperature value once an hour, but will have other records to indicate changes to wind or visibility before the next record with a temperature value,

Some of the value and error handling can, and probably should, be moved into pynws, but this requires quite a bit of refactoring the code in the sensor to continue to handle some edge cases.

I think the appropriate path forward at this point is to deal with this PR (#23647) first and then we can discuss how to improve the pynws library and refactor the sensor approach.

One note on testing the nws weather component:  you might want to look at the approach I used for the tests for the sensor:  I used the aio_client_mock library to have specific raw test responses provided back when requests are made down through pynws.  This allowed me to cover all of the code that handles data errors (or strange cases) just by using different data files.  It also allows for testing handling bad HTTP response codes:  I have encountered a number of events where api.weather.gov was returning HTTP errors for a period of time, so ensuring the stack would handle these without creating a lot of log entries, or worse, failing completely, seems important.

MatthewFlamm commented:
> I think the appropriate path forward at this point is to deal with this PR (#23647) first and then we can discuss how to improve the pynws library and refactor the sensor approach.

I would agree with this, and I would welcome PRs into pynws.  But I would caution that I do not see eye to eye with the philosophy of trying to jump through hoops to correct for deficiencies in the NWS API.  We should simply present the data.  I would support straightforward robustness measures like the metar code or using a look back over some time period (1hr?) for missing data.  I wouldn't support making complex decisions based on empirical observation.  I don't know if that changes your mind on adding functionality to pynws as a path forward

In my view, it may be that some stations are just not suitable for using in HA.  This should either be fixed on NWS side or someone needs to create a custom component to handle the super specific edge case of their station.  I haven't gotten any reports of problems from people using the custom component that mirrors this PR (I don't know how many people use it though), so I suspect it is a tiny minority that have these issues.

>  I used the aio_client_mock library to have specific raw test responses

I see this is also done on other some other integrations, but then we are testing both the pypi library and the home assistant code together.  Then you end up with tests that are testing the internals of the upstream library rather than the HA logic.  I would think that this runs counter to the principal of having a separate pypi library, but I will defer to the HA devs if this is acceptable.

MartinHjelmare commented:
Home Assistant wants to avoid making fake representations of device/api data in the primary integrations, and should present the data as is from the device/api, while maintaining the entity model that is in use. We have secondary integrations, eg template entities, that depend on the primary integrations, where the user can tailor the data to be represented.

If this isn't enough, custom components, are always a way out.

MatthewFlamm commented:
If @rewsssiestedbo is still OK with putting some of this functionality in pynws given my viewpoint, I suggest we move the conversation of the mechanics of how to handle these issues (stations providing only partial updates for example) into pynws.  I have a few ideas of how this could be done in a way that doesn't require edge-case detection.

@MartinHjelmare I will point out that weather platform does not have a template version.  I think this would be appreciated by people who have weather stations.  They could have observations provided by their ultra-local station, and the forecast provided by a different platform.

MatthewFlamm committed:
bump pynws version

MatthewFlamm committed:
update requirements_all

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,378 @@
+
+    if None in (latitude, longitude):
+        _LOGGER.error("Latitude/longitude not set in Home Assistant config")
+        return ConfigEntryNotReady
```
If it's a non recoverable error we should just `return` here after logging the error. We can't raise this error since this isn't a config entry.

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,378 @@
+            )
+        else:
+            self._forecast = self.nws.forecast
+        return
```
Not needed return. Please remove it.

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,378 @@
+            _LOGGER.error(
+                "Error updating forecast from station %s: %s", self.nws.station, status
+            )
+        else:
```
We can `return` here instead and outdent below.

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,378 @@
+        if self.observation:
+            temp_c = self.observation.get("temperature")
+        if temp_c:
+            return convert_temperature(temp_c, TEMP_CELSIUS, TEMP_FAHRENHEIT)
```
Why not report the temperature in celsius and let the core convert as needed?

MatthewFlamm replied:
The NWS API does not allow for a choosable unit system.  NWS provides the observation in C and the forecast in F.  Wonderful isn't it? Given the existence of unit conversion in HA, it is cleanest to do here.

MatthewFlamm replied:
I could move the conversion to the forecast if you'd prefer.

MartinHjelmare replied:
The temperature will be converted as needed here:
https://github.com/home-assistant/home-assistant/blob/f30e54f01e0460f08ff3d470ca45612eb9d860d8/homeassistant/components/weather/__init__.py#L126-L127

MartinHjelmare replied:
Ok, now I get it. Either we convert here or in forecast. Let's just keep it.

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,378 @@
+    websession = async_get_clientsession(hass)
+    # ID request as being from HA, pynws prepends the api_key in addition
+    api_key_ha = f"{api_key} homeassistant"
+    nws = SimpleNWS(latitude, longitude, api_key_ha, mode, websession)
```
We should check that the api key is valid, and if it's not, log an error and `return`.

MatthewFlamm replied:
The API can be any string, although an email address is requested for usage monitoring purposes, I presume.  This is in the docs.  I additionally used the request from HA, as this seems to meet the spirit of the request.  I'm okay taking the whole thing out too since it would still follow their guideline in the strict sense.

MartinHjelmare replied:
Ok.

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,378 @@
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the NWS weather platform."""
+    from pynws import SimpleNWS
```
This can be moved to the top of the module.

MatthewFlamm committed:
address comments

MatthewFlamm committed:
Merge branch 'add_nws' of github.com:MatthewFlamm/home-assistant into add_nws

MatthewFlamm committed:
move observation and forecast outside try-except-else

MatthewFlamm committed:
Revert "move observation and forecast outside try-except-else"

This reverts commit 53b78b32837b55b8a0b61de6192e846f6a486754.

MatthewFlamm committed:
remove else from update forecast block

MartinHjelmare commented:
Good! Now it's just the tests left to either adjust per my comment above, or remove them and exclude the integration in `.coveragerc`.

MatthewFlamm committed:
remove unneeded ConfigEntryNotReady import

MatthewFlamm commented:
Thanks!  I found the ipma integration had tests in pytest form that I can try working off of.  I will give it a shot, but if I'm making no progress, I will take the second route.

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,378 @@
+        self.observation = None
+        self._forecast = None
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
```
Should we adjust the default scan interval too, by setting `SCAN_INTERVAL` constant?

MatthewFlamm replied:
Do I simply just set the constant?  I see where it is set in the weather component, [here](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/weather/__init__.py#L42-L43) but I'm not sure how to modify it for the platform.

MartinHjelmare replied:
Yep. Just define it at the module level.

MatthewFlamm replied:
I made SCAN_INTERNAL 15 minutes, and the MIN_TIME_BETWEEN_UPDATES 5 minutes.  I haven't seen any stations that provide data in much more frequent intervals.

rewsssiestedbo replied:
I have seen cases where the time between two updates is less than 5 minutes - this is typically when wind gusts or visibility are changing rapidly, but it would not be a good idea to poll more frequently.

The normal update cycle for METAR stations is once per hour, plus updates for important changes between the hourly updates.   Unfortunately, these updates often only include the item or items that have changed, so if you only look at the latest update record, you can easily miss changes of other values.  This is why I had my sensor code retrieving a number of records and processing any that were newer than the previously processed update.  Without this, it is quite common to not get good values for some parameters for long periods of time:   unless the poll happens to catch the update with full data.  (Note that non-airport stations typically do not include a raw METAR record)

MatthewFlamm replied:
I captured this idea here.  https://github.com/MatthewFlamm/pynws/issues/15

MatthewFlamm committed:
add scan_interval, reduce min_time_between_updates

MatthewFlamm committed:
pytest tests

MatthewFlamm commented on homeassistant/components/nws/weather.py:
```diff
@@ -0,0 +1,379 @@
+    mode = config[CONF_MODE]
+
+    if None in (latitude, longitude):
+        _LOGGER.error("Latitude/longitude not set in Home Assistant config")
```
In testing, it seems there is a default latitude and longitude even when no latitude and longitude are specified anywhere, so is this if block unreachable?  Should I instead check whether config has both or neither? Is there a better way?

MartinHjelmare replied:
It's unreachable code. We can remove this check.

MatthewFlamm replied:
Thanks!  I rewrote this to become reachable.  PS I'm done with tests, and will push updates tomorrow hopefully.

MatthewFlamm committed:
lint test docstring

MartinHjelmare commented on homeassistant/components/nws/weather.py:
```diff
@@ -165,16 +164,18 @@ def convert_condition(time, weather):
 async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the NWS weather platform."""
 
+    if (config.get(CONF_LATITUDE) and not config.get(CONF_LONGITUDE)) or (
```
Latitude and longitude should be inclusive in the config schema. So we won't need this check after that change.

MartinHjelmare commented on tests/components/nws/test_weather.py:
```diff
@@ -0,0 +1,282 @@
+FORCURL = "https://api.weather.gov/points/{},{}/forecast"
+
+
+@asyncio.coroutine
```
Please use Python 3.5 async syntax with `async def` and `await`.

MatthewFlamm committed:
use async await

MatthewFlamm committed:
lat and lon inclusive in config

sofar commented:
I'm still really looking forward to using this from stock. I've been using it as a local integration for a while without issues and while I'm probably not running the latest, it's been well worth it - this beats many of the other services that require a subscription.

MatthewFlamm commented:
Is there anything else needed here?

@sofar If this doesn't move, I'll update the custom component repo to support HACS so it is easier to stay on top of versions.

MatthewFlamm commented:
At risk of setting this PR back even farther, I have a working config flow setup for this integration that basically just as asks for all the fields in the configuration.  I'm trying to figure out how to get some of the options into an option flow witha list of possible values.

For example,.I set lat/long in the config flow.  Then I set up my entity with the default weather station for the lat/lon.  The options flow ideally would get a list of nearly stations either using API call, or even better taking it from the entity directly, and provide a list to the user.  I cannot find any examples of this usage flow however.  I only see text forms and check boxes.  It doesn't even seem possible to dynamically update the entry names on the data flow form, which would be a hack method anyway.

MartinHjelmare commented:
Let's not add anything more to this PR. That can come in the next PR.

MartinHjelmare reviewed (approved):
Great!

MatthewFlamm commented:
Thank you for your reviewing time and support @MartinHjelmare, @fabaff, and @rewsssiestedbo.
